### PR TITLE
[data] Warn on excessive driver memory usage during shuffle ops

### LINKED ISSUE #40861: [data] Warn on high driver memory usage during shuffle ops
### Description

Excessive driver memory can occur from:
- intermediate block metadata
- intermediate block data (if they are small enough that they get stored in the driver's CoreWorker heap memory)

In both cases, the recommendation is to reduce the number of intermediate blocks by increasing block size or using push-based shuffle if not already used.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Warns the user if too much driver memory is used for:
- ObjectRef C++ metadata (estimated)
- Small objects that get stored in driver heap memory

## Related issue number

Closes #40861.

### CODE REVIEW

File: python/ray/data/_internal/execution/interfaces/task_context.py [hunk 1]
```diff
@@ -15,6 +20,11 @@ class TaskContext:
     # operator.
     task_idx: int
 
+    # If driver memory exceeds this threshold, warn the user. For now, this
+    # only applies to shuffle ops because most other ops are unlikely to use as
+    # much driver memory.
+    warn_on_driver_memory_usage_bytes: Optional[int] = None
```
Comment (c21) [inline python/ray/data/_internal/execution/interfaces/task_context.py:26]: Can we pass this argument through `DataContext` to `ExchangeTaskScheduler.execute(warn_on_driver_memory_usage_bytes=...)` directly?

`TaskContext` ideally should only contain the unique information per each task. Feeling it does not need to go through `TaskContext` in this case.

**Review by c21 — COMMENTED** [review]
Thanks @stephanie-wang. Change on data directory looks good to me, except one comment.

File: python/ray/data/_internal/execution/interfaces/task_context.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/data/_internal/execution/interfaces/task_context.py]: Thanks, moved to ExchangeTaskScheduler.

**Review by c21 — APPROVED** [review]
Thanks @stephanie-wang! The code change on Data part LGTM.

**Pull request merged**